File: labs/class-05-eda/sampler.py
#!/usr/bin/env python3
"""
Understanding Uncertainty - Proportions, one-hot encoding & the ECDF
sampler (standalone, stdlib only)
------------------------------------------------------------------
A sample proportion, its standard error, and the ECDF evaluated at a
threshold, computed from scratch. All three reduce to the same
formula: average a 0/1 indicator, then take sqrt(p_hat*(1-p_hat)/n)
for the standard error. Mirrors sampler.js and viz.js's EDA.* core.

Usage:
  python3 sampler.py proportion <label1,label2,...> <target>
  python3 sampler.py ecdf <v1,v2,...> <x>
  python3 sampler.py describe                 # this sample's key numbers

Reads the sample from data.json sitting next to this file.
"""
import sys, os, json, math


def proportion(labels, target):
    return sum(1 for v in labels if v == target) / len(labels)


def se(p, n):
    return math.sqrt(p * (1 - p) / n)


def ecdf_at(values, x):
    return sum(1 for v in values if v <= x) / len(values)


def load():
    here = os.path.dirname(os.path.abspath(__file__))
    with open(os.path.join(here, "data.json")) as f:
        return json.load(f)


def main():
    argv = sys.argv[1:]
    if not argv:
        sys.stderr.write("usage: python3 sampler.py [proportion|ecdf|describe] ...\n")
        sys.exit(1)
    cmd = argv[0]

    if cmd == "proportion":
        labels = argv[1].split(",")
        target = argv[2]
        p = proportion(labels, target)
        print(f"{p:.6f} {se(p, len(labels)):.6f}")
    elif cmd == "ecdf":
        values = [float(v) for v in argv[1].split(",")]
        x = float(argv[2])
        f = ecdf_at(values, x)
        print(f"{f:.6f} {se(f, len(values)):.6f}")
    elif cmd == "describe":
        data = load()
        n = data["n"]
        p = proportion(data["outcome"], 1)
        print(f"n:              {n}")
        print(f"p_hat(outcome=1): {p:.4f}  se: {se(p, n):.4f}")
        for label in ("1", "2", "3"):
            pl = proportion(data["triage"], label)
            print(f"p_hat(triage={label}):   {pl:.4f}  se: {se(pl, n):.4f}")
        for x in (60, 70, 80, 90, 100):
            f = ecdf_at(data["hr"], x)
            print(f"F_hat(HR<={x}):    {f:.4f}  se: {se(f, n):.4f}")
    else:
        sys.stderr.write("usage: python3 sampler.py [proportion|ecdf|describe] ...\n")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: labs/class-05-eda/sampler.js
#!/usr/bin/env node
/* Understanding Uncertainty - Proportions, one-hot encoding & the ECDF
 * sampler (standalone, pure JS, no dependencies)
 * ------------------------------------------------------------------
 * Mirrors sampler.py and viz.js's EDA.* core line-for-line.
 *
 * Usage (Node):
 *   node sampler.js proportion <label1,label2,...> <target>
 *   node sampler.js ecdf <v1,v2,...> <x>
 *
 * Reads the sample from data.json sitting next to this file (not
 * needed for the two commands above, which take their data inline).
 */
'use strict';

function proportion(labels, target) {
  return labels.filter(v => v === target).length / labels.length;
}
function se(p, n) { return Math.sqrt(p * (1 - p) / n); }
function ecdfAt(values, x) {
  return values.filter(v => v <= x).length / values.length;
}

function main() {
  const argv = process.argv.slice(2);
  if (!argv.length) {
    process.stderr.write('usage: node sampler.js [proportion|ecdf] ...\n');
    process.exit(1);
  }
  const cmd = argv[0];
  if (cmd === 'proportion') {
    const labels = argv[1].split(',');
    const target = argv[2];
    const p = proportion(labels, target);
    console.log(p.toFixed(6) + ' ' + se(p, labels.length).toFixed(6));
  } else if (cmd === 'ecdf') {
    const values = argv[1].split(',').map(parseFloat);
    const x = parseFloat(argv[2]);
    const f = ecdfAt(values, x);
    console.log(f.toFixed(6) + ' ' + se(f, values.length).toFixed(6));
  } else {
    process.stderr.write('usage: node sampler.js [proportion|ecdf] ...\n');
    process.exit(1);
  }
}

main();
